library/core/src/stream/stream/mod.rs

```diff
@@ -0,0 +1,129 @@
+    /// when the next value may be ready.
+    ///
+    /// - `Poll::Ready(Some(val))` means that the stream has successfully
+    /// produced a value, `val`, and may produce further values on subsequent
+    /// `poll_next` calls.
+    ///
+    /// - `Poll::Ready(None)` means that the stream has terminated, and
+    /// `poll_next` should not be invoked again.
+    ///
+    /// # Panics
+    ///
+    /// Once a stream has finished (returned `Ready(None)` from `poll_next`), calling its
+    /// `poll_next` method again may panic, block forever, or cause other kinds of
+    /// problems; the `Stream` trait places no requirements on the effects of
+    /// such a call. However, as the `poll_next` method is not marked `unsafe`,
+    /// Rust's usual rules apply: calls must never cause undefined behavior
+    /// (memory corruption, incorrect use of `unsafe` functions, or the like),
+    /// regardless of the stream's state.
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>>;
+
+    /// Returns the bounds on the remaining length of the stream.
+    ///
+    /// Specifically, `size_hint()` returns a tuple where the first element
+    /// is the lower bound, and the second element is the upper bound.
+    ///
+    /// The second half of the tuple that is returned is an [`Option`]`<`[`usize`]`>`.
+    /// A [`None`] here means that either there is no known upper bound, or the
+    /// upper bound is larger than [`usize`].
+    ///
+    /// # Implementation notes
+    ///
+    /// It is not enforced that a stream implementation yields the declared
+    /// number of elements. A buggy stream may yield less than the lower bound
+    /// or more than the upper bound of elements.
+    ///
+    /// `size_hint()` is primarily intended to be used for optimizations such as
+    /// reserving space for the elements of the stream, but must not be
+    /// trusted to e.g., omit bounds checks in unsafe code. An incorrect
+    /// implementation of `size_hint()` should not lead to memory safety
+    /// violations.
+    ///
+    /// That said, the implementation should provide a correct estimation,
+    /// because otherwise it would be a violation of the trait's protocol.
+    ///
+    /// The default implementation returns `(0, `[`None`]`)` which is correct for any
+    /// stream.
+    #[inline]
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        (0, None)
+    }
+
+    /// Advances the stream and returns a future which yields the next value.
+    ///
+    /// The returned future yields [`None`] when iteration is finished.
+    /// Individual stream implementations may choose to resume iteration, and so
+    /// calling `next()` again may or may not eventually start yielding
+    /// [`Some(Item)`] again at some point.
+    ///
+    /// [`Some(Item)`]: Some
+    fn next(&mut self) -> Next<'_, Self>
+    where
+        Self: Unpin,
+    {
```

Comment:
This makes the stream not object safe because it mentions `Self` in the return type.
```suggestion
    fn next(&mut self) -> Next<'_, Self>
    where
        Self: Sized + Unpin,
    {
```
Can we add a test that makes sure that the trait is object safe?

Reply:
Additionally, please be aware that this can cause problems with calling `next` on a `Pin<Box<dyn Stream>>`. We ran in to the same issue in https://github.com/tokio-rs/tokio/pull/3291 when trying to remove `AsyncReadExt` and `AsyncWriteExt` from Tokio.

For example, to call `.next()` on an `&mut dyn Stream`, you would have to do `(&mut the_stream_ref).next()` to use the impl of `Stream` for a `&mut dyn Stream`. This also requires separate impls of `Stream` for `&mut dyn Stream + Send` and `&mut dyn Stream + Send + Sync` and so on to be able to call the `next()` method at all on those kinds of trait objects.

Reply:
Here is a playground with a few variants: [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=ddef891c714e4b244c749a1129e54c22).

Reply:
@carllerche @Darksonn Good catch! It's not difficult to restore object safety, but not being able to directly call `.next` on `&mut dyn Stream` is bad...

Reply:
`Stream::next` has been removed from the RFC as per https://github.com/rust-lang/rfcs/pull/2996#issuecomment-757386206. This PR will be updated to reflect that.